### PR TITLE
Respond to events from other buffers, add caching, improve UX, fix bugs

### PR DESCRIPTION
**Pull request by b0o**
Hi! I've been maintaining a fork of twoslash-queries for about a month where I've made quite a few changes. Rather than breaking this into multiple smaller PRs, I'm going to submit this as one big PR. I'm happy to break this up into smaller PRs if that's preferred.

The most major change is that queries are updated when other files in the project are changed. This is nice when you have a query in one file that depends on another file, and you want to see the results of the query change when the other file changes.

In order to support this, I've added a cache to store the results of the queries so we can avoid deleting and re-creating extmarks if the query result hasn't changed. This prevents flickering that would happen when a buffer is modified.

Also, `:TwoslashQueriesInspect` now tries to guess the proper indent, so that when a file is formatted with prettier or another formatter, the query comment's indentation won't change and break the query.

In addition, there are a few other bug fixes, and I've formatted the code with stylua.

### CODE REVIEW

File: lua/twoslash-queries/init.lua [hunk 1]
```diff
@@ -140,8 +168,33 @@ end
...
+local clients = {}
+
+---@param client_id client_id
+local function update_types_for_client(client_id)
+  local cur = clients[client_id]
+  if not cur then
+    return
+  end
+  for _, bufnr in ipairs(cur.bufs) do
+    update_types(cur.client, bufnr)
```
Comment (marilari88) [inline lua/twoslash-queries/init.lua]: In some cases, the buffer is referred to as 'buf,' while in others it's referred to as 'buffer.' I'd call it 'buffer' for consistency.

File: lua/twoslash-queries/init.lua [hunk 2]
```diff
@@ -1,172 +1,343 @@
...
-local get_types = function(client, buffer_nr)
-	vim.api.nvim_buf_clear_namespace(buffer_nr, virtual_types_ns, 0, -1)
+local update_hover_text = function(client, buffer_nr, line, column, cb)
+  local target = vim.api.nvim_buf_get_lines(buffer_nr, line, line + 1, false)[1]
+  local position = { line = line - 2, character = column - 1 }
+  if not vim.api.nvim_buf_is_valid(buffer_nr) then
+    cb()
+    return
+  end
+  local params = { textDocument = vim.lsp.util.make_text_document_params(buffer_nr), position = position }
```
Comment (marilari88) [inline lua/twoslash-queries/init.lua]: we can move this assignation just before client.request

File: lua/twoslash-queries/init.lua [hunk 3]
```diff
@@ -1,172 +1,343 @@
...
+  if not client then
+    cb()
+    return
```
Comment (marilari88) [inline lua/twoslash-queries/init.lua]: we can join these conditions

Comment (marilari88) [reply lua/twoslash-queries/init.lua]: Hey @b0o Thank you for your awesome work on the project!  It took me a bit to digest all the changes, but that's probably on me 😂. I tested locally and everything works fine. I've left a couple of tiny comments.

After addressing these comments we can merge it as is for now and we can work on polishing it up after. I feel like we can improve the readability of the code but overall I think that Your improvements are solid!

Comment (b0o) [reply lua/twoslash-queries/init.lua]: Thanks for reviewing! I will go ahead and clean things up a bit, I agree the readability can use improvement.

**Comment by b0o** [comment]
Alright, pushed a pretty big refactor. I hope it's easier to follow now. Let me know what you think can be improved to make the code more readable.

**Comment by marilari88** [comment]
@b0o I allowed myself to change a few function namings. Thank you again for your huge work!

**Comment by b0o** [comment]
Awesome! Thank you.

**Pull request merged**